File: dec8.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

int width8, height8 = 0;

bool is_visible_fos(int forest[width8][height8], int x, int y) {
    int h = forest[x][y];

    //from right
    bool vis = true;
    for (int i = x + 1; i < width8; ++i)
        if(forest[i][y] >= h) vis = false;
    if(vis) return true;

    //from left
    vis = true;
    for (int i = x - 1; i >= 0; --i)
        if(forest[i][y] >= h) vis = false;
    if(vis) return true;

    //from bottom
    vis = true;
    for (int i = y + 1; i < height8; ++i)
        if(forest[x][i] >= h) vis = false;
    if(vis) return true;

    //from top
    vis = true;
    for (int i = y - 1; i >= 0; --i)
        if(forest[x][i] >= h) vis = false;
    if(vis) return true;

    return false;
}

int scenic_score(int forest[width8][height8], int x, int y) {
    int h = forest[x][y];
    int trees = 0;
    int score;
    //left
    for (int i = x + 1; i < width8; ++i) {
        trees++;
        if(forest[i][y] >= h) break;
    }
    score = trees;
    trees = 0;
    //right
    for (int i = x - 1; i >= 0; --i) {
        trees++;
        if(forest[i][y] >= h) break;
    }
    score *= trees;
    trees = 0;
    //down
    for (int i = y + 1; i < height8; ++i) {
        trees++;
        if(forest[x][i] >= h) break;
    }
    score *= trees;
    trees = 0;
    //up
    for (int i = y - 1; i >= 0; --i) {
        trees++;
        if(forest[x][i] >= h) break;
    }
    score *= trees;
    return score;
}

void sol8() {
    char const* const fileName = "C:\\Users\\Joachim\\CLionProjects\\AdventOfCode\\input8.txt";
    FILE* file = fopen(fileName, "r");

    char line[120];

    height8 = 0;
    //Get size
    while(fgets(line, 120, file)) {
        width8 = (int)strlen(line);
        height8 += 1;
    }

    //Load data
    int forest[width8][height8];
    fclose(file);
    file = fopen(fileName, "r");
    for (int i = 0; i < height8; ++i) {
        fgets(line, 120, file);
        for (int j = 0; j < width8; ++j) {
            forest[j][i] = line[j] - 48;
        }
    }

    //Solve
    int visible = 0;
    int best = 0;
    for (int y = 0; y < height8; ++y) {
        for (int x = 0; x < width8; ++x) {
            int score = scenic_score(forest, x, y);
            if (score > best) best = score;

            if(is_visible_fos(forest, x, y)) {
                visible++;
                //printf("%d", forest[x][y]);
            } //else printf(".");
        }
//        printf("\n");
    }

    printf("Part 1: %d\n", visible);
    printf("Part 2: %d\n", best);

    fclose(file);
}

File: dec8.pas
program dec8 (input, output, stderr);
var
    F: TextFile;
    line: AnsiString;
    score, best: LongInt;
    width, height, visible, i, j, x, y: Integer;
    forest: array[0..100] of array[0..100] of Integer;

function is_visible(x: integer; y: integer): Boolean;
var
    vis: Boolean;
    h: Integer;
begin
    h := forest[x][y];

    vis := True;
    for i := x + 1 to width - 1 do 
      if forest[i][y] >= h then vis := false;
    if vis then exit(True);

    vis := True;
    for i := x - 1 downto 0 do
      if forest[i][y] >= h then vis := false;
    if vis then exit(True);

    vis := True;
    for i := y + 1 to height - 1 do 
      if forest[x][i] >= h then vis := false;
    if vis then exit(True);

    vis := True;
    for i := y - 1 downto 0 do 
      if forest[x][i] >= h then vis := false;
    if vis then exit(True);

    is_visible := false;
end;

function scenic_score(x: integer; y: integer): LongInt;
var
    h: Integer;
    trees, scn: LongInt;
begin
    h := forest[x][y];
    trees := 0;
    scn := 0;

    for i := x + 1 to width - 1 do begin
      trees += 1;
      if forest[i][y] >= h then break;
    end;
    scn := trees;
    trees := 0;

    for i := x - 1 downto 0 do begin
      trees += 1;
      if forest[i][y] >= h then break;
    end;
    scn := trees * scn;
    trees := 0;

    for i := y + 1 to height - 1 do begin
      trees += 1;
      if forest[x][i] >= h then break;
    end;
    scn := trees * scn;
    trees := 0;

    for i := y - 1 downto 0 do begin
      trees += 1;
      if forest[x][i] >= h then break;
    end;
    scn := trees * scn;

    scenic_score := scn;
end;

//Program
begin
    Assign(F,'input8.txt');
    Reset(F);

    //Get size
    height := 0;
    while not Eof(F) do begin
      ReadLn(F, line);
      height += 1;
      width := length(line);
    end;
    Reset(F);

    //Load data
    for i := 1 to height do begin
      ReadLn(F, line);
      for j := 1 to width do begin
        forest[i - 1][j - 1] := ord(line[j]) - 48;
      end;
    end;

    //Solve
    visible := 0;
    best := 0;
    for x := 0 to height - 1 do begin
      for y := 0 to width - 1 do begin
        score := scenic_score(x, y);
        if score < 0 then WriteLn(score);
        if score > best then best := score;

        if is_visible(x, y) then begin 
          visible += 1;
          //Write(forest[x][y]);
        end// else Write('.');
      end;
      //WriteLn();
    end;

    WriteLn('Part 1: ', visible);
    WriteLn('Part 2: ', best);

    Close(F);
end.